docs/05_advanced_techniques/05_3_external_programs.ipynb

```diff
@@ -0,0 +1,287 @@
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Mit externen Programmen kommunizieren"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Python bietet die Möglichkeit an, andere Prozesse zu starten und mit ihnen zu kommunizieren."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Wenn du den Kommandozeilenbefehl kennst, um ein externes Programm auszuführen, dann ist es auch nicht schwer, dieses mit Python auszuführen.\n",
+    "\n",
+    "Möchtest du z.B. den Namen des aktuellen Users herausfinden, dann könntest du folgenden Code (ohne \"!\") in deinem Terminal eingeben:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "vscode": {
+     "languageId": "shellscript"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "!whoami"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Dieser Befehl gibt dir den aktuellen Benutzernamen zurück. Diesen Befehl kannst du auch via Python ausführen lassen, wobei die Ausgabe des Befehls direkt ge`print`ed wird:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import subprocess\n",
+    "\n",
+    "\n",
+    "subprocess.call(['whoami'])"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Möchtest du die Ausgabe in einer Python-Variablen speichern können, dann verwende eine andere Funktion:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_name = subprocess.check_output([\"whoami\"])\n",
+    "\n",
+    "print(f\"Your user name is {user_name}.\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In der Ausgabe stellst du fest, dass die Rückgabe ein Byte-Array statt einem String ist. Um `b'...'` loszuwerden, musst du die Rückgabe noch dekodieren:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_name = subprocess.check_output([\"whoami\"]).decode(\"utf-8\")\n",
+    "# Rufe noch `.strip()` auf, um die neue Zeile am Ende zu entfernen.\n",
+    "\n",
+    "print(f\"Your user name is {user_name}.\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Installierte Programme ausführen"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Mit Pythons `subprocess`-Modul können natürlich auch heruntergeladene installierte Programme aufgerufen werden.\n",
+    "\n",
+    "Z.B. kannst du auch Git-Befehle ausführen:"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
Damit dieser Code funktioniert, muss 'git' auf der Maschine installiert sein. Diese Vorraussetzung sollte erwähnt werden, damit nicht gewundert wird wenn beim Ausführen eine Fehlermeldung erscheint.

Reply:
Habe einen kleinen Kommentar ergänzt.

Jedoch könnte man im Moment das Einführungsprogramm ohne Git gar nicht klonen. Aber klar, je nach dem was in der Zukunft passiert, ist Git möglicherweise nicht installiert.